### PR TITLE
Ch16 Restructuring

### PR DESCRIPTION
**Pull request by Xmasreturns**
This PR started as a solution to #764 but ended up growing after some discussion. The old organization of Ch16 was confusing to read as a newcomer to rust. The new structure helps to introduce topics in order of increasing complexity and decreasing relevance, while leaving in some detail for more nuanced users:

16 - Error handling
16.1 - panic
16.2 - Option & unwrap
16.2.1 - Combinators: map
16.2.2 - Combinators: and_then
16.3 - Result
16.3.1 - map for Result
16.3.2 - aliases for Result
16.4 - Working with multiple error types
16.4.1 - Early returns
16.4.2 - try!
16.5 - Defining an error type
16.6 - Other uses of try!
16.7 - Boxing errors

Many sections were reworded for readability and grammar. Some were renamed, and some information was moved between sections. I apologize profusely for the poor commit structure here, as I'm positive it's somewhat hard to follow.

Shoutouts to @mdinger for helping with the Early returns example

### CODE REVIEW

File: examples/structure.json [hunk 1]
```diff
@@ -140,21 +140,22 @@
...
-        { "id": "unwrap", "title": "Option & unwrap", "children": null },
-        { "id": "map", "title": "map", "children": null },
-        { "id": "and_then", "title": "and_then", "children": null },
-        { "id": "result", "title": "Result", "children": [
-            { "id": "result_map", "title": "map for Result", "children": null },
-            { "id": "result_alias", "title": "aliases for Result", "children": null }
+        { "id": "panic", "title": "`panic`", "children": null },
+        { "id": "option_unwrap", "title": "`Option` & `unwrap`", "children": [
+            { "id": "map", "title": "Combinators: `map`", "children": null },
+            { "id": "and_then", "title": "Combinators: `and_then`", "children": null },
```
Comment (mdinger) [inline examples/structure.json]: `"children": null },` The trailing comma is a typo.

File: examples/error/input.md [hunk 2]
```diff
@@ -1,22 +1,9 @@
 Error handling is the process of handling the possibility of failure. For 
 example, failing to read a file and then continuing to use that *bad* input 
-would clearly be problematic. Error handling allows us to notice and handle 
-those errors in an explicit fashion, saving the rest of the program from 
-potential issues.
+would clearly be problematic. Noticing and managing those errors in an 
+explicit fashion saves the rest of the program from potential issues.
```
Comment (mdinger) [inline examples/error/input.md]: Maybe `potential pitfalls`. Nothing wrong here but I thought a slightly more specific word might be better. Not sure.

File: examples/error/input.md [hunk 3]
```diff
@@ -1,22 +1,9 @@
...
 
-The simplest error handling mechanism we will see is `panic`. It prints an 
-error message, starts unwinding the task, and usually exits the program. 
-Consider the following example:
+For additional information on error handling, see the section on error 
+handling in the [official book][book].
```
Comment (mdinger) [inline examples/error/input.md]: Nitpick: Maybe `For a more rigorous introduction...`. This is a good idea but we want people to at least try to read this section first before looking to the official version. The wording is typical for what you'd see at the end of a section, not at the beginning before anything is discussed.

File: examples/error/option_unwrap/and_then/input.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+then become confusing. That's where another combinator called `and_then()`, 
+known in some languages as flatmap, comes in.
+
+`and_then()` calls its function input with the wrapped value and returns the result. If the `Option` is `None`, then it returns `None` instead.
+
+In the following example, `cookable_v2()` results in an `Option<Food>`. 
+Using `map()` instead of `and_then()` would have given an 
+`Option<Option<Food>>`, which is an invalid type for `eat()`.
+
+{and_then.play}
```
Comment (mdinger) [inline examples/error/option_unwrap/and_then/input.md]: I was reading over this example (not the text) and I don't actually think it's that clear (clearly not your fault because you didn't even modify it). Well, it explains a use case for `and_then` fairly well but one thing I was expecting was a better comparison between `map` and `and_then`. They are both subtly different from each other and what that difference is wasn't very clear. [Here](https://is.gd/UscNeY) is a more obvious comparison.

Maybe an extra page should be added to compare the two? I think both examples are mainly fine except they aren't easily comparable. Opinions?

File: examples/error/result/input.md [hunk 5]
```diff
@@ -1,22 +1,33 @@
...
+describes possible *error* instead of possible *absence*.
+
+That is, `Result<T, E>` could have one of two outcomes:
 
 * `Ok<T>`: An element `T` was found
 * `Err<E>`: An error was found with element `E`
 
-Similar to `Option`, `Result` also contains the `unwrap()` method which yields the element
-`T` or calls `panic!()`. So far, this should seem similar to `Option`:
+By convention, the expected outcome is “Ok” while the unexpected outcome is “Err”.
```
Comment (mdinger) [inline examples/error/result/input.md]: I would do `Ok` and `Err` (wrapped in backticks), not `"Ok"` and `"Err"`

File: examples/error/result/input.md [hunk 6]
```diff
@@ -1,22 +1,33 @@
...
+to `panic` on. Additionally, the `panic` exits our program and provides an 
+unpleasant error message.
 
-[`Result`][result]
+To improve the quality of our error message, we should be more specific 
+about the return type and consider explicitly handling the error.
 
+[option]: http://doc.rust-lang.org/std/option/enum.Option.html
 [result]: http://doc.rust-lang.org/std/result/enum.Result.html
+[parse]: https://doc.rust-lang.org/std/primitive.str.html#method.parse
```
Comment (mdinger) [inline examples/error/result/input.md]: I like how this page has been reworked. I like it a lot better.

File: examples/error/result/result_map/input.md [hunk 7]
```diff
@@ -1,24 +1,19 @@
...
-for [`i32`][i32]. As a result, the `Err` type is specified as [`ParseIntError`][parse_int_error].
+Panicking in the previous example gave us an unhelpful error message.
+To avoid that, we need to be more specific about the return type. There, the 
+regular element is of type `i32`.
 
-In the example below, note that using the straightforward `match` statement leads to 
-more cumbersome code. As it turns out, the `map` method we used with `Option` 
-is also implemented for `Result`.
+To determine the `Err` type, we look to 
+`parse()`, which is implemented with the [`FromStr`][from_str] trait for
```
Comment (mdinger) [inline examples/error/result/result_map/input.md]: I would link to [`parse()`](https://doc.rust-lang.org/std/primitive.str.html#method.parse) when mentioning it.

File: examples/error/multiple_error_types/input.md [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+To do so, we convert both the `Option` and `Result` into `Result`s, and 
+then map their errors to the same type:
+
+{multiple_error_types.play}
+
+In the next section, we'll see an alternate method of explicitly handling these errors.
+
+### See Also:
+
+[`Option::ok_or`][okor], [Result::map_err][maperr]
```
Comment (mdinger) [inline examples/error/multiple_error_types/input.md]: Add backticks around `Result::map_err`

File: examples/structure.json (hunk 1 above)
Comment (Xmasreturns) [inline examples/structure.json]: Good eye, this one was bugging me.

File: examples/error/result/result_map/input.md (hunk 7 above)
Comment (Xmasreturns) [inline examples/error/result/result_map/input.md]: Agreed, good change.

File: examples/error/result/input.md (hunk 6 above)
Comment (Xmasreturns) [inline examples/error/result/input.md]: Thanks, I really appreciate you taking the time to look over it!

File: examples/error/result/input.md (hunk 5 above)
Comment (Xmasreturns) [inline examples/error/result/input.md]: Agreed, good change.

File: examples/error/multiple_error_types/input.md (hunk 8 above)
Comment (Xmasreturns) [inline examples/error/multiple_error_types/input.md]: Good eye, I missed this one.

**Comment by mdinger** [comment]
The `16.4.2 Introducing try!` example doesn't fit the structure anymore as written. It is worded as if the previous example had deeper and deeper nesting which now can be avoided. It also switches to a different example which has not been seen before. I think I'd reword it based on https://is.gd/kAXzoU as an example (or something similar) to retain the original example and continue pushing towards proper error handling.

The followup page continues towards more proper error handling so it seems like a decent intermediary.

File: examples/error/define_error_type/define_error_type.rs [hunk 9]
```diff
@@ -4,31 +4,28 @@ use std::fmt;
...
+// Define our error types. These may be customized for our error handling cases. 
+// Now we will be able to write our own errors, defer to an underlying error
+// implementation, or do something in between.
 enum DoubleError {
     // We don't require any extra info to detail this error.
     EmptyVec,
-    // We will defer to the parse error implementation for their error. Supplying extra
-    // info would require adding more data to the type.
+    // We will defer to the parse error implementation for their error.
+    // Supplying extra info requires adding more data to the type.
```
Comment (mdinger) [inline examples/error/define_error_type/define_error_type.rs]: I would note https://github.com/rust-lang/rust-by-example/issues/782 for completeness but wouldn't worry about it here. This is a comprehensive section rewrite which should need to be concerned with something like that. It could be added later.

File: examples/error/reenter_try/input.md [hunk 10]
```diff
@@ -1,38 +1,44 @@
...
+could be elided. Alas, because `and_then` is not sufficiently flexible, it 
+cannot. However, we can instead use `try!`.
+
+`try!` was previously explained as either `unwrap` or `return Err(err)`. 
+This is only mostly true. It actually means `unwrap` or 
+`return Err(From::from(err))`. Since `From::from` is a conversion utility 
+between different types, this means that if you `try!` where the error is 
+convertible to the return type, it will convert automatically.
+
+Here, we rewrite the previous example using `try!` As a result, the
```
Comment (mdinger) [inline examples/error/reenter_try/input.md]: I think there should be a period here...kind weird case.

File: examples/error/input.md (hunk 2 above)
Comment (Xmasreturns) [inline examples/error/input.md]: I understand the sentiment here, and a rewording was in order.

File: examples/error/input.md (hunk 3 above)
Comment (Xmasreturns) [inline examples/error/input.md]: This is a good point, and will be changed.

File: examples/error/reenter_try/input.md (hunk 10 above)
Comment (Xmasreturns) [inline examples/error/reenter_try/input.md]: Typo on my part. This will come in the next commit.

**Comment by mdinger** [comment]
In general, I think this is really good. Most of the updates are grammar fixing, chapter restructuring, and title changes. The structural changes make sense as do the title changes. Thanks so much for your efforts! I know large section rewriting like this takes a lot of time and effort!

File: examples/error/option_unwrap/and_then/input.md (hunk 4 above)
Comment (Xmasreturns) [inline examples/error/option_unwrap/and_then/input.md]: I can see where you're going with the direct comparison there. I think it makes sense to simply add a small unrunnable code box to show off part of your example. In my opinion, the `input.md` text as it stands is pretty much sufficient for explaining the difference if paired with some code comparison. What do you think?

Comment (mdinger) [inline examples/error/option_unwrap/and_then/input.md]: Yes. That seems like a good idea. The problem is the text is correct but abstract. Giving something concrete to look at should be quite helpful.

**Comment by mdinger** [comment]
@Xmasreturns I meant to look this over again last weekend but I didn't get around to it. Hopefully this weekend but I might be busy. We'll see.

**Comment by steveklabnik** [comment]
I like it. Thank you!
